File: Providers/WorldWeatherOnline/WWOProvider.cs
using System;
using System.IO;
using System.Net;
using System.Xml.XPath;

namespace WeatherAPI.Providers.WorldWeatherOnline {
	
	internal class WWOProvider : WeatherProvider, IWeather {
		
		private const string WWO_API_FORMAT = "xml";
		private const string WWO_API_URL = "http://free.worldweatheronline.com/feed/weather.ashx?q={0}&format={1}&num_of_days=1&key={2}";
		private const string WWO_XPATH_HEADER = "//data/current_condition/{0}";
		
		private XPathDocument _xpath;
		private string _wwo_api_key;
		
		public WWOProvider () : base() {
			if (IsAvailable()) {
				_wwo_api_key = _dllConfig.AppSettings.Settings["WORLD_WEATER_ONLINE_API_KEY"].Value;
			}
		}
		
		public override bool IsAvailable() {
			return _dllConfig.AppSettings.Settings["WORLD_WEATER_ONLINE_API_KEY"] != null &&
					!String.IsNullOrEmpty(_dllConfig.AppSettings.Settings["WORLD_WEATER_ONLINE_API_KEY"].Value);
		}
		
		public override void Update() {
			string url = String.Format(WWO_API_URL, Location, WWO_API_FORMAT, _wwo_api_key);
			HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
			
			HttpWebResponse response = (HttpWebResponse)request.GetResponse();
			string xml = new StreamReader(response.GetResponseStream()).ReadToEnd();
			
			_xpath = new XPathDocument(new StringReader(xml));
		}
		
		public override bool Supports(LocationSource source) { 
			return true;
		}
		
		public double DegreesCelcius {
			get {
				string xpath = String.Format(WWO_XPATH_HEADER, "temp_C/text()");
				object val = _xpath.CreateNavigator().SelectSingleNode(xpath);
				
				return Double.Parse(val.ToString());
			}
		}

		public double DegressFahrienhiet {
			get {
				string xpath = String.Format(WWO_XPATH_HEADER, "temp_F/text()");
				object val = _xpath.CreateNavigator().SelectSingleNode(xpath);
				
				return Double.Parse(val.ToString());
			}
		}

		public double WindSpeedMPH {
			get {
				string xpath = String.Format(WWO_XPATH_HEADER, "windspeedMiles/text()");
				object val = _xpath.CreateNavigator().SelectSingleNode(xpath);
				
				return Double.Parse(val.ToString());
			}
		}
		
		public double WindSpeedKPH {
			get {
				string xpath = String.Format(WWO_XPATH_HEADER, "windspeedKmph/text()");
				object val = _xpath.CreateNavigator().SelectSingleNode(xpath);
				
				return Double.Parse(val.ToString());
			}
		}
		
		public Direction WindDirection {
			get {
				string xpath = String.Format(WWO_XPATH_HEADER, "winddir16Point/text()");
				object val = _xpath.CreateNavigator().SelectSingleNode(xpath);
				
				return (Direction)Enum.Parse(typeof(Direction), val.ToString());
			}
		}

		public double CloudCover {
			get {
				string xpath = String.Format(WWO_XPATH_HEADER, "cloudcover/text()");
				object val = _xpath.CreateNavigator().SelectSingleNode(xpath);
				
				return Double.Parse(val.ToString());
			}
		}
		
		public double Percipitation {
			get {
				string xpath = String.Format(WWO_XPATH_HEADER, "precipMM/text()");
				object val = _xpath.CreateNavigator().SelectSingleNode(xpath);
				
				return Double.Parse(val.ToString());
			}
		}

		public double Humidity { 
			get {
				string xpath = String.Format(WWO_XPATH_HEADER, "humidity/text()");
				object val = _xpath.CreateNavigator().SelectSingleNode(xpath);
				
				return Double.Parse(val.ToString());
			}
		}
		
		public WeatherCondition Conditions {
			get {
				string xpath = String.Format(WWO_XPATH_HEADER, "weatherCode/text()");
				object val = _xpath.CreateNavigator().SelectSingleNode(xpath);
				
				WWOWeatherCode code = (WWOWeatherCode)(int.Parse(val.ToString()));
				
				WeatherCondition currentCondition = translateConditions(code);
				
				return currentCondition;
			}
		}
		
		protected WeatherCondition translateConditions(WWOWeatherCode code) {
			switch (code) {
				case WWOWeatherCode.Blizzard:
					return WeatherCondition.Blizzard | WeatherCondition.Snow;
				case WWOWeatherCode.BlowingSnow:
					return WeatherCondition.Blowing | WeatherCondition.Snow;
				case WWOWeatherCode.ClearSunny:
					return WeatherCondition.Clear;
				case WWOWeatherCode.Cloudy:
					return WeatherCondition.Cloudy;
				case WWOWeatherCode.Fog:
					return WeatherCondition.Fog;
				case WWOWeatherCode.FreezingDrizzle:
					return WeatherCondition.Freezing | WeatherCondition.Drizzle;
				case WWOWeatherCode.FreezingFog:
					return WeatherCondition.Freezing | WeatherCondition.Fog;
				case WWOWeatherCode.HeavyFreezingDrizzle:
					return WeatherCondition.Heavy | WeatherCondition.Freezing | WeatherCondition.Drizzle;
				case WWOWeatherCode.HeavyRain:
					return WeatherCondition.Heavy | WeatherCondition.Rain;
				case WWOWeatherCode.HeavySnow:
					return WeatherCondition.Heavy | WeatherCondition.Snow;
				case WWOWeatherCode.IcePellets:
					return WeatherCondition.Ice | WeatherCondition.Pellets;
				case WWOWeatherCode.LightDrizzle:
					return WeatherCondition.Light | WeatherCondition.Drizzle;
				case WWOWeatherCode.LightFreezingRain:
					return WeatherCondition.Light | WeatherCondition.Freezing | WeatherCondition.Rain;
				case WWOWeatherCode.LightIcePellets:
					return WeatherCondition.Light | WeatherCondition.Ice | WeatherCondition.Pellets;
				case WWOWeatherCode.LightRain:
					return WeatherCondition.Light | WeatherCondition.Rain;
				case WWOWeatherCode.LightRainShower:
					return WeatherCondition.Light | WeatherCondition.Rain | WeatherCondition.Showers;
				case WWOWeatherCode.LightSleet:
					return WeatherCondition.Light | WeatherCondition.Sleet;
				case WWOWeatherCode.LightSleetShowers:
					return WeatherCondition.Light | WeatherCondition.Sleet | WeatherCondition.Showers;
				case WWOWeatherCode.LightSnow:
					return WeatherCondition.Light | WeatherCondition.Snow;
				case WWOWeatherCode.LightSnowShowers:
					return WeatherCondition.Light | WeatherCondition.Snow | WeatherCondition.Showers;
				case WWOWeatherCode.Mist:
					return WeatherCondition.Mist;
				case WWOWeatherCode.ModerateOrHeavyFreezingRain:
					return WeatherCondition.Moderate | WeatherCondition.Heavy | WeatherCondition.Freezing | WeatherCondition.Rain;
				case WWOWeatherCode.ModerateOrHeavyIcePellets:
					return WeatherCondition.Moderate | WeatherCondition.Heavy | WeatherCondition.Ice | WeatherCondition.Pellets;
				case WWOWeatherCode.ModerateOrHeavyRainAndThunder:
					return WeatherCondition.Moderate | WeatherCondition.Heavy | WeatherCondition.Rain | WeatherCondition.Thunder;
				case WWOWeatherCode.ModerateOrHeavyRainShowers:
					return WeatherCondition.Moderate | WeatherCondition.Heavy | WeatherCondition.Rain | WeatherCondition.Showers;
				case WWOWeatherCode.ModerateOrHeavySleet:
					return WeatherCondition.Moderate | WeatherCondition.Heavy | WeatherCondition.Sleet;
				case WWOWeatherCode.ModerateOrHeavySleetshowers:
					return WeatherCondition.Moderate | WeatherCondition.Heavy | WeatherCondition.Sleet | WeatherCondition.Showers;
				case WWOWeatherCode.ModerateOrHeavySnowShowers:
					return WeatherCondition.Moderate | WeatherCondition.Heavy | WeatherCondition.Snow | WeatherCondition.Showers;
				case WWOWeatherCode.ModerateOrHeavySnowWithThunder:
					return WeatherCondition.Moderate | WeatherCondition.Heavy | WeatherCondition.Snow | WeatherCondition.Thunder;
				case WWOWeatherCode.ModerateRain:
					return WeatherCondition.Moderate | WeatherCondition.Rain;
				case WWOWeatherCode.ModerateSnow:
					return WeatherCondition.Moderate | WeatherCondition.Snow;
				case WWOWeatherCode.Overcast:
					return WeatherCondition.Overcast;
				case WWOWeatherCode.PartlyCloudy:
					return WeatherCondition.PartlyCloudy;
				case WWOWeatherCode.PatchyFreezingDrizzleNearby:
					return WeatherCondition.Patchy | WeatherCondition.Freezing | WeatherCondition.Drizzle;
				case WWOWeatherCode.PatchyHeavyRain:
					return WeatherCondition.Patchy | WeatherCondition.Heavy | WeatherCondition.Rain;
				case WWOWeatherCode.PatchyHeavySnow:
					return WeatherCondition.Patchy | WeatherCondition.Heavy | WeatherCondition.Snow;
				case WWOWeatherCode.PatchyLightDrizzle:
					return WeatherCondition.Patchy | WeatherCondition.Light | WeatherCondition.Drizzle;
				case WWOWeatherCode.PatchyLightRain:
					return WeatherCondition.Patchy | WeatherCondition.Light | WeatherCondition.Rain;
				case WWOWeatherCode.PatchyLightRainAndThunder:
					return WeatherCondition.Patchy | WeatherCondition.Light | WeatherCondition.Rain | WeatherCondition.Thunder;
				case WWOWeatherCode.PatchyLightSnow:
					return WeatherCondition.Patchy | WeatherCondition.Light | WeatherCondition.Snow;
				case WWOWeatherCode.PatchyLightSnowWithThunder:
					return WeatherCondition.Patchy | WeatherCondition.Light | WeatherCondition.Snow | WeatherCondition.Thunder;
				case WWOWeatherCode.PatchyModerateRain:
					return WeatherCondition.Patchy | WeatherCondition.Moderate | WeatherCondition.Rain;
				case WWOWeatherCode.PatchyModerateSnow:
					return WeatherCondition.Patchy | WeatherCondition.Moderate | WeatherCondition.Snow;
				case WWOWeatherCode.PatchyRainNearby:
					return WeatherCondition.Patchy | WeatherCondition.Rain;
				case WWOWeatherCode.PatchySleetNearby:
					return WeatherCondition.Patchy | WeatherCondition.Sleet;
				case WWOWeatherCode.PatchySnowNearby:
					return WeatherCondition.Patchy | WeatherCondition.Snow;
				case WWOWeatherCode.ThunderyOutbreaksInNearby:
					return WeatherCondition.Thunder;
				case WWOWeatherCode.TorrentialRainShowers:
					return WeatherCondition.Torrential | WeatherCondition.Rain | WeatherCondition.Showers;
				default:
					return WeatherCondition.Clear;
			}
		}
	}
}

File: src/main/java/WeatherAPI/Providers/WorldWeatherOnline/WWOProvider.java
package WeatherAPI.Providers.WorldWeatherOnline;
	
import WeatherAPI.Direction;
import WeatherAPI.IWeather;
import WeatherAPI.Providers.LocationSource;
import WeatherAPI.Providers.WeatherProvider;
import WeatherAPI.WeatherCondition;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.EnumSet;

public class WWOProvider extends WeatherProvider implements IWeather {

	private static final String WWO_API_FORMAT = "xml";
	private static final String WWO_API_URL = "http://free.worldweatheronline.com/feed/weather.ashx?q=%s&format=%s&num_of_days=1&key=%s";
	private static final String WWO_XPATH_HEADER = "//data/current_condition/%s";

	private XMLReader _reader;
	private String _wwo_api_key;

	public WWOProvider () {
		if (IsAvailable()) {
			_wwo_api_key = PROPERTIES.getProperty("WORLD_WEATHER_ONLINE_API_KEY");
			_reader = new XMLReader();
		}
	}

	public boolean IsAvailable() {
		return PROPERTIES.containsKey("WORLD_WEATHER_ONLINE_API_KEY") &&
			   (PROPERTIES.getProperty("WORLD_WEATHER_ONLINE_API_KEY") != null ||
				!PROPERTIES.getProperty("WORLD_WEATHER_ONLINE_API_KEY").equals(""));
	}

	public void Update() {
		URL url;
		HttpURLConnection conn = null;
		InputStream stream = null;

		try {
			url = new URL(String.format(WWO_API_URL, getLocation(), WWO_API_FORMAT, _wwo_api_key));
			conn = (HttpURLConnection)url.openConnection();
			stream = conn.getInputStream();
			StringBuilder sb = new StringBuilder();
			int data = stream.read();

			while (data != -1) {
				sb.append((char)data);
				data = stream.read();
			}

			_reader.load(sb.toString());
			_lastUpdate = System.currentTimeMillis();
		}
		catch (Exception e) {
			e.printStackTrace(System.err);
		}
		finally {
			try { if (stream != null) stream.close(); }
			catch (IOException e) { e.printStackTrace(System.err); }
			finally { if (conn != null ) conn.disconnect(); }
		}
	}

	public boolean Supports(LocationSource source) { 
		return true;
	}

	public double getDegreesCelcius() {
		if (!isFresh()) { Update(); }
		String xpath = String.format(WWO_XPATH_HEADER, "temp_C/text()");
		Object val = _reader.read(xpath);

		return Double.parseDouble(val.toString());
	}

	public double getDegressFahrienhiet() {
		if (!isFresh()) { Update(); }
		String xpath = String.format(WWO_XPATH_HEADER, "temp_F/text()");
		Object val = _reader.read(xpath);

		return Double.parseDouble(val.toString());
	}

	public double getWindSpeedMPH() {
		if (!isFresh()) { Update(); }
		String xpath = String.format(WWO_XPATH_HEADER, "windspeedMiles/text()");
		Object val = _reader.read(xpath);

		return Double.parseDouble(val.toString());
	}

	public double getWindSpeedKPH() {
		if (!isFresh()) { Update(); }
		String xpath = String.format(WWO_XPATH_HEADER, "windspeedKmph/text()");
		Object val = _reader.read(xpath);

		return Double.parseDouble(val.toString());
	}

	public Direction getWindDirection() {
		if (!isFresh()) { Update(); }
		String xpath = String.format(WWO_XPATH_HEADER, "winddir16Point/text()");
		Object val = _reader.read(xpath);

		return Direction.valueOf(val.toString());
	}

	public double getCloudCover() {
		if (!isFresh()) { Update(); }
		String xpath = String.format(WWO_XPATH_HEADER, "cloudcover/text()");
		Object val = _reader.read(xpath);

		return Double.parseDouble(val.toString());
	}

	public double getPercipitation() {
		if (!isFresh()) { Update(); }
		String xpath = String.format(WWO_XPATH_HEADER, "precipMM/text()");
		Object val = _reader.read(xpath);

		return Double.parseDouble(val.toString());
	}

	public double getHumidity() {
		if (!isFresh()) { Update(); }
		String xpath = String.format(WWO_XPATH_HEADER, "humidity/text()");
		Object val = _reader.read(xpath);

		return Double.parseDouble(val.toString());
	}

	public EnumSet<WeatherCondition> getConditions() {
		if (!isFresh()) { Update(); }
		String xpath = String.format(WWO_XPATH_HEADER, "weatherCode/text()");
		Object val = _reader.read(xpath);

		WWOWeatherCode code = WWOWeatherCode.findByCode(Integer.parseInt(val.toString()));

		EnumSet<WeatherCondition> currentCondition = translateConditions(code);

		return currentCondition;
	}

	protected EnumSet<WeatherCondition> translateConditions(WWOWeatherCode code) {
		switch (code) {
			case Blizzard:
				return EnumSet.of(WeatherCondition.Blizzard, WeatherCondition.Snow);
			case BlowingSnow:
				return EnumSet.of(WeatherCondition.Blowing, WeatherCondition.Snow);
			case ClearSunny:
				return EnumSet.of(WeatherCondition.Clear);
			case Cloudy:
				return EnumSet.of(WeatherCondition.Cloudy);
			case Fog:
				return EnumSet.of(WeatherCondition.Fog);
			case FreezingDrizzle:
				return EnumSet.of(WeatherCondition.Freezing, WeatherCondition.Drizzle);
			case FreezingFog:
				return EnumSet.of(WeatherCondition.Freezing, WeatherCondition.Fog);
			case HeavyFreezingDrizzle:
				return EnumSet.of(WeatherCondition.Heavy, WeatherCondition.Freezing, WeatherCondition.Drizzle);
			case HeavyRain:
				return EnumSet.of(WeatherCondition.Heavy, WeatherCondition.Rain);
			case HeavySnow:
				return EnumSet.of(WeatherCondition.Heavy, WeatherCondition.Snow);
			case IcePellets:
				return EnumSet.of(WeatherCondition.Ice, WeatherCondition.Pellets);
			case LightDrizzle:
				return EnumSet.of(WeatherCondition.Light, WeatherCondition.Drizzle);
			case LightFreezingRain:
				return EnumSet.of(WeatherCondition.Light, WeatherCondition.Freezing, WeatherCondition.Rain);
			case LightIcePellets:
				return EnumSet.of(WeatherCondition.Light, WeatherCondition.Ice, WeatherCondition.Pellets);
			case LightRain:
				return EnumSet.of(WeatherCondition.Light, WeatherCondition.Rain);
			case LightRainShower:
				return EnumSet.of(WeatherCondition.Light, WeatherCondition.Rain, WeatherCondition.Showers);
			case LightSleet:
				return EnumSet.of(WeatherCondition.Light, WeatherCondition.Sleet);
			case LightSleetShowers:
				return EnumSet.of(WeatherCondition.Light, WeatherCondition.Sleet, WeatherCondition.Showers);
			case LightSnow:
				return EnumSet.of(WeatherCondition.Light, WeatherCondition.Snow);
			case LightSnowShowers:
				return EnumSet.of(WeatherCondition.Light, WeatherCondition.Snow, WeatherCondition.Showers);
			case Mist:
				return EnumSet.of(WeatherCondition.Mist);
			case ModerateOrHeavyFreezingRain:
				return EnumSet.of(WeatherCondition.Moderate, WeatherCondition.Heavy, WeatherCondition.Freezing, WeatherCondition.Rain);
			case ModerateOrHeavyIcePellets:
				return EnumSet.of(WeatherCondition.Moderate, WeatherCondition.Heavy, WeatherCondition.Ice, WeatherCondition.Pellets);
			case ModerateOrHeavyRainAndThunder:
				return EnumSet.of(WeatherCondition.Moderate, WeatherCondition.Heavy, WeatherCondition.Rain, WeatherCondition.Thunder);
			case ModerateOrHeavyRainShowers:
				return EnumSet.of(WeatherCondition.Moderate, WeatherCondition.Heavy, WeatherCondition.Rain, WeatherCondition.Showers);
			case ModerateOrHeavySleet:
				return EnumSet.of(WeatherCondition.Moderate, WeatherCondition.Heavy, WeatherCondition.Sleet);
			case ModerateOrHeavySleetshowers:
				return EnumSet.of(WeatherCondition.Moderate, WeatherCondition.Heavy, WeatherCondition.Sleet, WeatherCondition.Showers);
			case ModerateOrHeavySnowShowers:
				return EnumSet.of(WeatherCondition.Moderate, WeatherCondition.Heavy, WeatherCondition.Snow, WeatherCondition.Showers);
			case ModerateOrHeavySnowWithThunder:
				return EnumSet.of(WeatherCondition.Moderate, WeatherCondition.Heavy, WeatherCondition.Snow, WeatherCondition.Thunder);
			case ModerateRain:
				return EnumSet.of(WeatherCondition.Moderate, WeatherCondition.Rain);
			case ModerateSnow:
				return EnumSet.of(WeatherCondition.Moderate, WeatherCondition.Snow);
			case Overcast:
				return EnumSet.of(WeatherCondition.Overcast);
			case PartlyCloudy:
				return EnumSet.of(WeatherCondition.PartlyCloudy);
			case PatchyFreezingDrizzleNearby:
				return EnumSet.of(WeatherCondition.Patchy, WeatherCondition.Freezing, WeatherCondition.Drizzle);
			case PatchyHeavyRain:
				return EnumSet.of(WeatherCondition.Patchy, WeatherCondition.Heavy, WeatherCondition.Rain);
			case PatchyHeavySnow:
				return EnumSet.of(WeatherCondition.Patchy, WeatherCondition.Heavy, WeatherCondition.Snow);
			case PatchyLightDrizzle:
				return EnumSet.of(WeatherCondition.Patchy, WeatherCondition.Light, WeatherCondition.Drizzle);
			case PatchyLightRain:
				return EnumSet.of(WeatherCondition.Patchy, WeatherCondition.Light, WeatherCondition.Rain);
			case PatchyLightRainAndThunder:
				return EnumSet.of(WeatherCondition.Patchy, WeatherCondition.Light, WeatherCondition.Rain, WeatherCondition.Thunder);
			case PatchyLightSnow:
				return EnumSet.of(WeatherCondition.Patchy, WeatherCondition.Light, WeatherCondition.Snow);
			case PatchyLightSnowWithThunder:
				return EnumSet.of(WeatherCondition.Patchy, WeatherCondition.Light, WeatherCondition.Snow, WeatherCondition.Thunder);
			case PatchyModerateRain:
				return EnumSet.of(WeatherCondition.Patchy, WeatherCondition.Moderate, WeatherCondition.Rain);
			case PatchyModerateSnow:
				return EnumSet.of(WeatherCondition.Patchy, WeatherCondition.Moderate, WeatherCondition.Snow);
			case PatchyRainNearby:
				return EnumSet.of(WeatherCondition.Patchy, WeatherCondition.Rain);
			case PatchySleetNearby:
				return EnumSet.of(WeatherCondition.Patchy, WeatherCondition.Sleet);
			case PatchySnowNearby:
				return EnumSet.of(WeatherCondition.Patchy, WeatherCondition.Snow);
			case ThunderyOutbreaksInNearby:
				return EnumSet.of(WeatherCondition.Thunder);
			case TorrentialRainShowers:
				return EnumSet.of(WeatherCondition.Torrential, WeatherCondition.Rain, WeatherCondition.Showers);
			default:
				return EnumSet.of(WeatherCondition.Clear);
		}
	}
}